### ISSUE TITLE
Issuing new tokens by capacity validators

### ISSUE THREAD

**Issue by robvanmieghem** [threefoldfoundation/tfchain#155]
It should be possible to create new tokens based on the provided capacity.

Since there is no capacity registration/validation mechanism yet, the capacity validation will be done by designated "capacity validators".
In more technical terms: a multisig of 9 out of twelve hardcoded addresses is required to sign a "token issuance" transaction. The arbitrary data field  can be used to justify the token issuance.

**Comment by GlenDC** [comment]
Do you want to do this using a normal transaction? If so, we would need to look how we specify this exceptional rule. We would also have to check how we allow Rivine to allow chains like tfchain to customise when coin outputs can be created without matching inputs (rather than only allowing this for genesis tokens).

Or do do this using a new Transaction Type, only within tfchain. It would make some things easier, but perhaps it is overkill for a feature like this, given that we already allow unfunded outputs for the genesis block.

**Comment by robvanmieghem** [comment]
I thought a new transaction type would be easier, it would also make it very clear that it is a coin creation transaction. Unless there is a good reason to do it differently off course.

**Comment by GlenDC** [comment]
> I thought a new transaction type would be easier.

Yes

> it would also make it very clear that it is a coin creation transaction

Yes

> Unless there is a good reason to do it differently off course.

I don't really see one.

**Comment by GlenDC** [comment]
```javascript
{
    "version": 128,
    "data": {
        // slice of PublicKeySignaturePairs,
        // where only the hardcoded keys are allowed
        "keysignaturepairs": [...],
        "coinoutputs": [...],
        // or perhaps you want to name it differently,
        // and perhaps you even want it as a structure,
        // which defines the backing behind these created coins
        "arbitrarydata": [...],
    },
}
```

I think we can keep it as small as this. The transaction type
would have custom validation.

Whether we want to keep it as simplistic as this is the question.
This proposal is simple because:

* it keeps the transaction data to the bare minimum;
* it hardcodes the keys that can co-sign such a coin-creation transaction;

It might however be useful to be able to change who is authorized to do so in the future,
right now that would rely an update to the code, with a block-height check as safety measure,
but what if we would instead allow (via a transaction) that the already authorized keys
can co-sign to define the keys authorized from that point on. This would require that the first
generation is hardcoded, but at least from than on it could change without needing a code update.

**Comment by robvanmieghem** [comment]
I agree and the dynamic update through the chain is a good idea as well, keeps us from forcing everyone to update.

**Comment by GlenDC** [comment]
@robvanmieghem here are the proposals (in JSON-encoded format) of how I see the 2 transaction types, which will allow the creation of coins as well as the definition of (a) new minter(s):

```javascript
{
    "version": 128, // TransactionMinterDefinition
    "data": {
        // fulfillment to be fulfilled in order to ensure only the
        // currently defined minter(s) can define the new minter(s)
        //
        // by default it is the hardcoded condition that has to be fulfilled, embedded in the source code,
        // if this hardcoded condition is overwritten using transaction of type #128,
        // it will use the newest registered condition as the condition to be fulfilled
        "fulfillment": {},
        // new condition that defines who become(s) the new minter(s),
        // and thus defines who can create coins as well as update who is/are the current minter(s)
        "condition": {},
    },
}

{
    "version": 129, // TransactionCoinCreation
    "data": {
        // fulfillment to be fulfilled in order to ensure only the
        // currently defined minter(s) can create coins
        //
        // by default it is the hardcoded condition that has to be fulfilled, embedded in the source code,
        // if this hardcoded condition is overwritten using transaction of type #128,
        // it will use the newest registered condition as the condition to be fulfilled
        "fulfillment": {},
        // coins to be created and locked by the given conditions
        "coinoutputs": [...],
        // maximum 83 bytes,
        // can contain anything, and is used to proof
        // the source of the coin creation
        "arbitrarydata": [...],
    },
}
```

**Comment by GlenDC** [comment]
As agreed the hardcoded initial condition to be fulfilled when sending a transaction of type `128` or `129`, is a MultiSignatureCondition with the 3 addresses and a MinimumSignatureCount of `3`. One owned by me, another by @robvanmieghem and another by @LeeSmet. Please provide an address you own and have secured.

Once the actual (12?) addresses and MinimumSignatureCount (9?) are known, we would than create a transaction of type `128`, by fulfilling with our 3 signatures, and defining the new MultiSignatureCondition.

From that point on, only those 12 (?) addresses can create coins —by signing with at least 9 (?) out of 12 (?)— as well as redefine the "mint" condition.

As you can see in this flow, the suggested approach is simple —as I said before— but it does work. On top of that it will not require much work at all to implement, given that all pieces of the puzzle are already developed. Which will allow us to spend more time on testing.

**Comment by GlenDC** [comment]
Allowing the miners to be defined at any point, using the `TransactionMinterDefinition` (128), does imply that we somehow keep track of the last minter definition (being it the hardcoded one, or the newest `TransactionMinterDefinition`. The most logical place for such a feature would be in the consensus, given that it keeps track of the blockchain state, and is also the place where all blocks are validated already. 

Given that Rivine (Go) is however not really a library, but more of a configurable binary, where you can hot-plug a couple of things, but where most things are still hardcoded, it makes it very hard to add this feature to there. There are 2 fast options:

+ Either we store the content in a bolt db only created and used by `tfchaind`, this would allow us to store any consensus change we want, including this;
+ We allow rivine implementations to define what the consensusdb stores: this is pretty weird though, and not sure how that would all fit together. On top of that the question would be how to access it safely. Seems like very bad design to me;

Where the first option is the only real option I see, with the current Rivine repo as it is.

It does really show, that the Rivine design as it is smells, when desired to be used as library. Rivine itself should have just been about the protocol, and optionally outside that core some utility code and extra building blocks. But the real blockchain should have been implemented directly in tfchain, as that would allow it to store whatever info it has to store.

Perhaps I am seeing it all wrong here, but adding the `TransactionMinterDefinition` to tfchain does seem trickier than foreseen. Nothing changes for `TransactionCoinCreation` that still stays simple (if we remove the feature where the minters can change who the minters are at least).

**Comment by GlenDC** [comment]
More complex JSON-encoded example of a CoinCreationTransaction (`129`):

```json
{
	"version": 129,
	"data": {
		"mintfulfillment": {
			"type": 3,
			"data": {
				"pairs": [
					{
						"publickey": "ed25519:def123def123def123def123def123def123def123def123def123def123def1",
						"signature": "ef12345ef12345ef12345ef12345ef12345ef12345ef12345ef12345ef12345ef12345ef12345ef12345ef12345ef12345ef12345ef12345ef12345ef12345ef"
					},
					{
						"publickey": "ed25519:ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff",
						"signature": "abcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefab"
					}
				]
			}
		},
		"coinoutputs": [
			{
				"value": "10000221",
				"condition": {
					"type": 4,
					"data": {
						"unlockhashes": [
							"01e89843e4b8231a01ba18b254d530110364432aafab8206bea72e5a20eaa55f70b1ccc65e2105",
							"01a6a6c5584b2bfbd08738996cd7930831f958b9a5ed1595525236e861c1a0dc353bdcf54be7d8"
						],
						"minimumsignaturecount": 2
					}
				}
			},
			{
				"value": "5000000000000000",
				"condition": {
					"type": 3,
					"data": {
						"locktime": 42,
						"condition": {
							"type": 1,
							"data": {
								"unlockhash": "01e89843e4b8231a01ba18b254d530110364432aafab8206bea72e5a20eaa55f70b1ccc65e2105"
							}
						}
					}
				}
			}
		],
		"minerfees": [
			"3000000",
			"1230000000"
		],
		"arbitrarydata": "ZGF0YQ=="
	}
}
```

**Comment by GlenDC** [comment]
Closed by b5164e7edbb6365c785a4b1ea157701a915be5a5.
